src/curl_jolokia.c

```diff
@@ -0,0 +1,1038 @@
+		value_list_t vl = VALUE_LIST_INIT;
+
+		db->curr_attribute = &db->attributepool[i];
+		llentry_t *cfgptr = llist_search_custom(db->match_this_bean->attributes,
+					     cjo_cfg_compare_attribute, db);
+		if (cfgptr == NULL) {
+			char attribute[db->curr_attribute->json_name_len + 1];
+
+			memcpy(attribute, db->curr_attribute->json_name,
+			       db->curr_attribute->json_name_len);
+			attribute[db->curr_attribute->json_name_len] = '\0';
+			ERROR("curl_jolokia plugin: failed to locate attribute [%s:\"%s\"]",
+			      db->match_this_bean->bean_name, attribute);
+
+			continue; /* we don't know this! */
+		}
+		cjo_attribute_t *curr_attribute = cfgptr->value;
+
+		/* Create a null-terminated version of the string. */
+		int ds_type = cjo_get_type(curr_attribute);
+		if (ds_type < 0) {
+			char attribute[db->curr_attribute->json_name_len + 1];
+
+			memcpy(attribute, db->curr_attribute->json_name,
+			       db->curr_attribute->json_name_len);
+			attribute[db->curr_attribute->json_name_len] = '\0';
+
+			ERROR("curl_jolokia plugin: failed to map type for [%s:%s:%s]",
+			      db->match_this_bean->bean_name, attribute, curr_attribute->type);
+			continue;
+		}
+
+		value_t ret_value = {0};
+
+		char buffer[db->max_value_len + 1];
+		memcpy(buffer, db->curr_attribute->json_value,
+		       db->curr_attribute->json_value_len);
+
+		buffer[db->curr_attribute->json_value_len] = '\0';
+
+		if (parse_value(buffer, &ret_value, ds_type) != 0) {
+			char attribute[db->curr_attribute->json_name_len + 1];
+
+			memcpy(attribute, db->curr_attribute->json_name,
+			       db->curr_attribute->json_name_len);
+			attribute[db->curr_attribute->json_name_len] = '\0';
+
+			WARNING("curl_jolokia plugin: Unable to parse number: [%s:%s:\"%s\"]",
+				db->match_this_bean->bean_name, attribute, buffer);
+			continue;
+		}
+
+		vl.values = &ret_value;
+		vl.values_len = 1;
+
+		char *host;
+		if ((db->host == NULL) || (strcmp("", db->host) == 0) ||
+		    (strcmp(CJO_DEFAULT_HOST, db->host) == 0))
+			host = "";
+		else
+			host = db->host;
+
+		sstrncpy(vl.type_instance, curr_attribute->attribute_name,
+			 sizeof(vl.type_instance));
+		sstrncpy(vl.plugin_instance, db->match_this_bean->bean_name,
+			 sizeof(vl.type_instance));
+
+		sstrncpy(vl.host, host, sizeof(vl.host));
+		if (db->match_this_bean->mbean_namespace == NULL)
+			sstrncpy(vl.plugin, "jolokia", sizeof(vl.plugin));
+		else
+			sstrncpy(vl.plugin, db->match_this_bean->mbean_namespace,
+				 sizeof(vl.plugin));
+		sstrncpy(vl.type, curr_attribute->type, sizeof(vl.type));
+
+		plugin_dispatch_values(&vl);
+	}
+
+	/* flush values */
+	db->match_this_bean = NULL;
+	db->attribute_pool_used = 0;
+
+	memset(db->attributepool, 0,
+	       sizeof(*db->attributepool) * db->max_attribute_count);
+
+} /* }}} int cjo_submit */
+
+/* yajl callbacks */
+#define CJO_CB_ABORT 0
+#define CJO_CB_CONTINUE 1
+
+static int cjo_cb_string(void *ctx, const unsigned char *val, size_t len) {
```

Comment:
Could you please add a comment that demonstrates what the expected JSON looks like? That would help me understand this logic tremendously.

Reply:
Their documentation explains it: https://jolokia.org/reference/html/protocol.html as previously noted, this thing is basically a dumbed down curl_json, which takes a pre-configured json post document, and based on the rest of the configuration tries to match the tree structure of the jolokia reply.

Reply:
The streaming nature of yail makes it a bit more complicated. If there was a way to address stuff like you do in https://stedolan.github.io/jq/ it would be easy ;-)